### ISSUE TITLE
Request to add  [kubernetes.io/ceph-secrets-protection] annotation to protect Ceph secrets from mistaken deletion

### ISSUE THREAD

**Issue by dotunolafunmiloye** [rook/rook#8369]
* Feature Request

Recently, the secrets and services for a Ceph cluster were deleted in error.  Yes, this should not have happened, but stuff happens. I was wondering if it possible to request a feature to add some protection to these secrets so it's not so easy to delete them.
For example, kubernetes pvc's already use the `Finalizers:    [kubernetes.io/pvc-protection]`  annotation as a form of protection and this could be adopted to safeguard Ceph secrets.

**What should the feature do:**
A protection annotation of the form `Finalizers:    [kubernetes.io/ceph-secrets-protection]`  will protect  Ceph secrets from mistaken deletion. If a deletion is really needed,  the protection can easily be removed, similar to removing pvc protection before deletion

**What is use case behind this feature:**
The use case here, is that Ceph secrets will not be so easy to delete. The addition of an annotation protection adds a stopgap for sanity. As with kubernetes pvc's the protected secrets can be deleted by removing the annotation.

**Environment**:

This is  could be applicable to all environments.

**Comment by leseb** [comment]
Which Secret/Service was deleted? Typically, if they are the operator should re-generate them.

**Comment by dotunolafunmiloye** [comment]
Unfortunately, all secrets were deleted. Yes, the  operator re-created a number of secrets automatically. The following : osd's, mon,  mgr an rgw were all re-created based on entries in  keyring files on the running pods. I  have compared the # of secrets to that of another stable cluster and looks good. Just need guidance on sequence of restarting pods and what to look out for. I have opened  https://github.com/rook/rook/issues/8369; hopefully to prevent secrets from being deleted so easily. Thank you.

**Comment by travisn** [comment]
As much as I hesitate to add finalizers to resources, adding finalizers to the critical secrets and configmaps would make sense since the operator cannot regenerate them in a disaster. All the resources that can be regenerated would not need a finalizer.
@leseb @BlaineEXE thoughts?

**Comment by leseb** [comment]
I'm not a big fan of adding finalizers everywhere, even if the request is to "prevent accidental deletion", still the resources will be deleted eventually. Plus, if the Rook operator fails to re-create a resource we just need to know whether this is a bug or just impossible (underlying Ceph resource is gone).

The request feels more like the former. So in your case @dotunolafunmiloye finalizers would not help much.

**Comment by travisn** [comment]
@leseb I'm thinking we should do more to protect a couple of critical configmaps and secrets, but not try to protect all resources. If those critical configmaps and secrets are deleted re-generated, the cluster loses the critical ceph keys and mon endpoints. With the finalizers, if those resources are deleted accidentally, they could still be restored with a process similar to [this](https://rook.github.io/docs/rook/v1.7/ceph-disaster-recovery.html#restoring-crds-after-deletion).

**Comment by dotunolafunmiloye** [comment]
> I'm not a big fan of adding finalizers everywhere, even if the request is to "prevent accidental deletion", still the resources will be deleted eventually. Plus, if the Rook operator fails to re-create a resource we just need to know whether this is a bug or just impossible (underlying Ceph resource is gone).
> 
> The request feels more like the former. So in your case @dotunolafunmiloye finalizers would not help much.

I think it will at least remove the element of big error/mistake. Finalizers stop the secret/cm from  being deleted in stealth. Before eventual deletion, the secret yaml file could be cloned  to a different filename, and a `kubectl create -f <filename>` used to re-create it saving the day, or the cluster.

**Comment by leseb** [comment]
@travisn @dotunolafunmiloye  agreed, but is it a disaster scenario or just a mistake? If it's a mistake secret/cm will be re-generated by the Operator right? If it's a disaster then even the finalizers won't help.

I'm still failing to understand why we need finalizers unless someone can tell me precisely which resource is not re-generated upon manual (mistake) deletion :-). Then I'll be ok with adding finalizers.

**Comment by dotunolafunmiloye** [comment]
In my case, it was a mistake. All ceph secrets were deleted.  Ceph recreated the core secrets. But all *keyring, rook-csi-cephfs* and rook-csi-rbd* were not re-created, leading to a long journey trying to restore and recover the cluster.  Thank you.

**Comment by travisn** [comment]
@leseb If the following resources are deleted, the operator gets into a bad state after it regenerates different keys and fsid that doesn't match the cluster. This metadata could likely be found from the running ceph daemons as described in the [disaster recovery guide](https://rook.github.io/docs/rook/v1.7/ceph-disaster-recovery.html). If it's recoverable we could also argue that it's not worth adding a finalizer since it should be so rare an occurrence.

Secrets:
- `rook-ceph-mon`: The admin username, secret, and fsid
ConfigMaps:
- `rook-ceph-mon-endpoints`: The expected mon endpoints

**Comment by dotunolafunmiloye** [comment]
Very true. I experienced these issues for the first time. The operator did re-generate some secrets as explained before. I was not knowledgeable enough to understand that a new fsid was auto-generated. I was able to re-create all secrets that were not automatically re-created by using the contents of keyring files on running pods. The `disaster recovery` steps were very useful to clear up many errors.  If a new fsid was generated, that may have been my issue. I attempted to re-use the old fsid. Just could not get rid of this error ```2021-07-25 17:56:34.704538 I | ceph-spec: ceph-file-controller: CephCluster "rook-ceph" found but skipping reconcile since ceph health is &{"HEALTH_ERR" map["error":{"Urgent" "failed to get status. . timed out: exit status 1"}] "2021-07-25T16:49:33Z" "2021-07-23T02:09:28Z" "HEALTH_OK"}
2021-07-25 17:56:34.704568 D | ceph-spec: "ceph-block-pool-controller": CephCluster resource "rook-ceph" found in namespace "rook-ceph"```.  For now, I have adopted the use of  `finalizers: kubernetes.io/secrets-protection` on our clusters to avoid a repeat. Thanks for your assistance.

**Comment by leseb** [comment]
Alright, thanks @travisn and @dotunolafunmiloye for the details. That sounds reasonable to add finalizers to `rook-ceph-mon` and `rook-ceph-mon-endpoints`.

**Comment by dotunolafunmiloye** [comment]
Yoohoo!! Thanks a million. Thanks for your service to the unseen millions and keeping Ceph strong!
